### PR TITLE
Added regression testcases for multiple memory server

### PR DESCRIPTION
**Pull request by rajakrishi**
Added testcases for these API:
  - Create Region
  - Destroy region
  - Resize
  - Permission Check (For dataitem & Region)

### CODE REVIEW

**Comment by rajakrishi** [comment]
Can one of the admins verify this patch?

File: test/reg-test/fam-api-reg/fam_control_path_reg_test.cpp [hunk 1]
```diff
@@ -0,0 +1,214 @@
+/*
+ * fam_copy_reg_test.cpp
+ * Copyright (c) 2019 Hewlett Packard Enterprise Development, LP. All rights
```
Comment (mashoodk) [inline test/reg-test/fam-api-reg/fam_control_path_reg_test.cpp:3]: Update copyright to 2020

File: test/reg-test/fam-api-reg/fam_control_path_reg_test.cpp [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+    int ret;
+    ::testing::InitGoogleTest(&argc, argv);
+
+    for (int i = 1; i < argc; ++i) {
+        printf("arg %2d = %s\n", i, (argv[i]));
+    }
+    if (argc == 2) {
+       NUM_MM_ITERATIONS = atoi(argv[1]);
+    }
+#if !defined(SHM) && defined(MEMSERVER_PROFILE)
```
Comment (mashoodk) [inline test/reg-test/fam-api-reg/fam_control_path_reg_test.cpp:191]: This #ifdef is not needed i think. Is this there in other regression tests ?

File: test/reg-test/fam-api-reg/fam_control_path_reg_test.cpp [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+
+int main(int argc, char **argv) {
+    int ret;
+    ::testing::InitGoogleTest(&argc, argv);
+
+    for (int i = 1; i < argc; ++i) {
+        printf("arg %2d = %s\n", i, (argv[i]));
+    }
+    if (argc == 2) {
+       NUM_MM_ITERATIONS = atoi(argv[1]);
```
Comment (mashoodk) [inline test/reg-test/fam-api-reg/fam_control_path_reg_test.cpp:189]: What is the default value of NUM_MM_ITERATIONS

File: test/reg-test/fam-api-reg/fam_control_path_reg_test.cpp [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+
+
+
+int main(int argc, char **argv) {
+    int ret;
+    ::testing::InitGoogleTest(&argc, argv);
+
+    for (int i = 1; i < argc; ++i) {
+        printf("arg %2d = %s\n", i, (argv[i]));
+    }
```
Comment (mashoodk) [inline test/reg-test/fam-api-reg/fam_control_path_reg_test.cpp:187]: if we are passing command line to regression test, we will need to update cmake file with args. Also we need to supply default values..

**Review by mashoodk — CHANGES_REQUESTED** [review]

File: test/reg-test/fam-api-reg/fam_control_path_reg_test.cpp [hunk 5]
```diff
@@ -0,0 +1,209 @@
+/*
+ * fam_copy_reg_test.cpp
```
Comment (mashoodk) [inline test/reg-test/fam-api-reg/fam_control_path_reg_test.cpp:2]: File name in comment is wrong..

File: test/reg-test/fam-api-reg/fam_control_path_reg_test.cpp [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+ */
+#include <fam/fam_exception.h>
+#include <gtest/gtest.h>
+#include <iostream>
+#include <stdio.h>
+#include <string.h>
+
+#include <fam/fam.h>
+
+#include "cis/fam_cis_client.h"
```
Comment (mashoodk) [inline test/reg-test/fam-api-reg/fam_control_path_reg_test.cpp:38]: You can remove this header file.

File: test/reg-test/fam-api-reg/fam_control_path_reg_test.cpp [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+
+int *myPE;
+fam *my_fam;
+Fam_CIS_Client *cis;
+Fam_Options fam_opts;
+int NUM_MM_ITERATIONS = 100;
+
+
+// Test case 1 - fam_copy and fam_copy_wait test (success).
+TEST(FamDataPath, CreateDestroyRegion) {
```
Comment (mashoodk) [inline test/reg-test/fam-api-reg/fam_control_path_reg_test.cpp:60]: This is not FamDataPath. May be you can name it as FamControlPath

File: test/reg-test/fam-api-reg/CMakeLists.txt [hunk 8]
```diff
@@ -31,7 +31,7 @@ add_fam_test(fam_options_reg_test)
 if (${TEST_RUNTIME} STREQUAL "PMI2")
 	add_fam_test(fam_runtime_pmi2_reg_test)
 	add_fam_test(fam_runtime_pmi2_abort_reg_test)
-elseif(${TEST_RUNTIME} STREQUAL "PMIX") 
+elseif(${TEST_RUNTIME} STREQUAL "PMIX")
```
Comment (mashoodk) [inline test/reg-test/fam-api-reg/CMakeLists.txt:34]: Shouldnt have these changes.

File: test/reg-test/fam-api-reg/fam_control_path_reg_test.cpp [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+#include <stdio.h>
+#include <string.h>
+
+#include <fam/fam.h>
+
+#include "cis/fam_cis_client.h"
+#include "common/fam_test_config.h"
+#define BIG_REGION_SIZE 214748364
+#define BLOCK_SIZE 1048576
+#define RESIZE_REGION_SIZE 1048576
```
Comment (mashoodk) [inline test/reg-test/fam-api-reg/fam_control_path_reg_test.cpp:42]: keep RESIZE_REGION_SIZE as same as BIG_REGION_SIZE

File: test/reg-test/fam-api-reg/fam_control_path_reg_test.cpp [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+
+ EXPECT_NO_THROW(my_fam->fam_barrier_all());
+ Fam_Stat *info = (Fam_Stat *)malloc(sizeof(Fam_Stat));
+ for (int i = 0; i < NUM_MM_ITERATIONS; i++) {
+     EXPECT_NO_THROW(my_fam->fam_stat(descLocal, info));
+     EXPECT_NO_THROW(my_fam->fam_resize_region(descLocal, info->size * i));
+ }
+ free(info);
+ EXPECT_NO_THROW(my_fam->fam_barrier_all());
+ // EXPECT_NO_THROW(rpc->generate_profile());
```
Comment (mashoodk) [inline test/reg-test/fam-api-reg/fam_control_path_reg_test.cpp:123]: remove commented lines..

File: test/reg-test/fam-api-reg/fam_control_path_reg_test.cpp [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+TEST(FamDataPath, FamResizeRegion) {
+ Fam_Region_Descriptor *descLocal;
+ const char *testRegionLocal = get_uniq_str("testLocal", my_fam);
+ EXPECT_NO_THROW(descLocal = my_fam->fam_create_region(
+                     testRegionLocal, RESIZE_REGION_SIZE, 0777, RAID1));
+ EXPECT_NE((void *)NULL, descLocal);
+
+ EXPECT_NO_THROW(my_fam->fam_barrier_all());
+ Fam_Stat *info = (Fam_Stat *)malloc(sizeof(Fam_Stat));
+ for (int i = 0; i < NUM_MM_ITERATIONS; i++) {
```
Comment (mashoodk) [inline test/reg-test/fam-api-reg/fam_control_path_reg_test.cpp:117]: I think we need to rethink this resize in a loop logic - We do not need to resize NUM_MM_ITERATIONS times. Couple of iterations should be enough

File: test/reg-test/fam-api-reg/fam_control_path_reg_test.cpp [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+                     testRegionLocal, BIG_REGION_SIZE, 0777, RAID1));
+ EXPECT_NE((void *)NULL, descLocal);
+
+ // Allocating data items
+ EXPECT_NO_THROW(itemLocal = my_fam->fam_allocate(
+                     firstItemLocal, DATA_ITEM_SIZE, 0777, descLocal));
+ EXPECT_NE((void *)NULL, itemLocal);
+ mode_t perm;
+ EXPECT_NO_THROW(my_fam->fam_barrier_all());
+ for (int i = 0; i < NUM_MM_ITERATIONS; i++) {
```
Comment (mashoodk) [inline test/reg-test/fam-api-reg/fam_control_path_reg_test.cpp:143]: We shouldnt change permission for lots of iterations. We can allocate many dataitems and change its permission.
Basically this loop can be for allocate - and then change permission. 

Say 
for (int i = 0; i < NUM_MM_ITERATIONS; i++) {
allocate()
change_permission()
 }

File: test/reg-test/fam-api-reg/fam_control_path_reg_test.cpp [hunk 13]
```diff
@@ -0,0 +1,209 @@
...
+TEST(FamDataPath, RegionChangePermission) {
+ Fam_Region_Descriptor *descLocal;
+ const char *testRegionLocal = get_uniq_str("testLocal", my_fam);
+ EXPECT_NO_THROW(descLocal = my_fam->fam_create_region(
+                     testRegionLocal, BIG_REGION_SIZE, 0777, RAID1));
+ EXPECT_NE((void *)NULL, descLocal);
+
+ EXPECT_NO_THROW(my_fam->fam_barrier_all());
+ mode_t perm;
+ for (int i = 0; i < NUM_MM_ITERATIONS; i++) {
```
Comment (mashoodk) [inline test/reg-test/fam-api-reg/fam_control_path_reg_test.cpp:167]: Same here - The loop around change permission in microbenchmark was to take performance measurement. We do not need that loop here.

**Review by mashoodk — CHANGES_REQUESTED** [review]

File: test/microbench/fam-api-mb/fam_microbenchmark_allocator.cpp [hunk 14]
```diff
@@ -319,8 +319,13 @@ TEST(FamChangePermissions, DataItemChangePermission) {
     Fam_Descriptor *itemLocal;
     const char *testRegionLocal = get_uniq_str("testLocal", my_fam);
     const char *firstItemLocal = get_uniq_str("firstLocal", my_fam);
-    EXPECT_NO_THROW(descLocal = my_fam->fam_create_region(
+    try {
+         EXPECT_NO_THROW(descLocal = my_fam->fam_lookup_region(testRegionLocal));
```
Comment (mashoodk) [inline test/microbench/fam-api-mb/fam_microbenchmark_allocator.cpp:323]: i think you will need to remove EXPECT_NO_THROW around fam_lookup_region, test case will fail - seeing an expect no throw and try catch around is not useful.

File: test/microbench/fam-api-mb/fam_microbenchmark_allocator.cpp [hunk 15]
```diff
@@ -319,8 +319,13 @@ TEST(FamChangePermissions, DataItemChangePermission) {
...
+    } catch (Fam_Exception &e) {
+        EXPECT_NO_THROW(descLocal = my_fam->fam_create_region(
                         testRegionLocal, BIG_REGION_SIZE, 0777, RAID1));
+        //cout << "fam region failed" << endl;
```
Comment (mashoodk) [inline test/microbench/fam-api-mb/fam_microbenchmark_allocator.cpp:327]: you can remove this line

File: test/reg-test/fam-api-reg/fam_control_path_reg_test.cpp [hunk 16]
```diff
@@ -0,0 +1,209 @@
...
+
+#include "common/fam_test_config.h"
+#define BIG_REGION_SIZE 214748364
+#define BLOCK_SIZE 1048576
+#define RESIZE_REGION_SIZE 214748364
+#define NAME_BUFF_SIZE 255
+#define DATA_ITEM_SIZE 1048576
+
+
+//#define MESSAGE_SIZE 12
```
Comment (mashoodk) [inline test/reg-test/fam-api-reg/fam_control_path_reg_test.cpp:46]: remove commented code.

**Review by mashoodk — CHANGES_REQUESTED** [review]

**Review by mashoodk — APPROVED** [review]

**Pull request merged**